cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1327,29 +1346,54 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             let src_ty = ctx.input_ty(insn, 0);
             let dst_ty = ctx.output_ty(insn, 0);
 
+            // Sextend requires a sign-extended move, but all the other opcodes are simply a move
+            // from a zero-extended source. Here is why this works, in each case:
+            //
+            // - Bint: Bool-to-int. We always represent a bool as a 0 or 1, so we merely need to
+            // zero-extend here.
+            //
+            // - Breduce, Bextend: changing width of a boolean. We represent a bool as a 0 or 1, so
+            // again, this is a zero-extend / no-op.
+            //
+            // - Ireduce: changing width of an integer. Smaller ints are stored with undefined
+            // high-order bits, so we can simply do a copy.
+
+            if src_ty == types::I32 && dst_ty == types::I64 && op != Opcode::Sextend {
+                // As a particular x86 peephole optimization, all the ALU opcodes on 32-bits will
+                // zero-extend the upper 32-bits, so we can even not generate a zero-extended move
+                // in this case.
+                if let Some(_) = matches_input_any(
+                    ctx,
+                    inputs[0],
+                    &[
```

Comment:
All 32-bit shifts on Intel (even signed right) will produce bits[63:32] == 0 too.  Also, probably a more common case .. a 32-bit load also has this property.

Reply:
Added a TODO.